cpp/src/arrow/util/crc32.cc

```diff
@@ -0,0 +1,961 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+/*
+ * crc32.c
+ *
+ * Copyright (c) 2011-2015 Stephan Brumme. All rights reserved.
```

Comment:
FYI: there was a crc32 implementation based on hardware accelerator in Arrow. Removed due to not used. This lookup table approach is not optimal for modern x86 and Arm CPUs, can be improved later.
https://github.com/apache/arrow/blob/2a8e37d4ac69d8fc26479e30228bf926ee35f6c5/cpp/src/arrow/util/hash_util.h

Reply:
As mentioned here: https://github.com/apache/arrow/pull/14351#discussion_r1004717971
Though I modify some code, I put it still under util. But I think the code you refered looks better

Reply:
@cyb70289  I asked for the code to be put under `util` because we want to take ownership of it in order to add potential performance improvements.

As for HW-accelerated CRC32, great, but should make sure we are talking about the same CRC32 parameters.

Reply:
Also, we need unit tests for the CRC32 functionality using well-known test vectors (or generated from a well-known implementation, such as [Python's](https://docs.python.org/3/library/binascii.html#binascii.crc32)).